packages/patterns/src/keys/keycollection-operators.js

```diff
@@ -0,0 +1,226 @@
+// @ts-check
+import {
+  assertRankSorted,
+  compareAntiRank,
+  makeFullOrderComparatorKit,
+  sortByRank,
+} from '@endo/marshal';
+import { makeIterator, makeArrayIterator } from '../utils.js';
+
+/** @typedef {import('@endo/marshal').RankCompare} RankCompare */
+/** @typedef {import('../types').KeyComparison} KeyComparison */
+/** @typedef {import('../types').KeyCompare} KeyCompare */
+/** @typedef {import('../types').FullCompare} FullCompare */
+/** @typedef {import('../types').KeyCollection} KeyCollection */
+
+const { quote: q, Fail } = assert;
+
+/**
+ * Refines a sequence of entries that is already sorted over its keys by the
+ * `rankCompare` preorder, where there may be internal runs tied for the same
+ * rank, into an iterable that resolves those ties using `fullCompare`.
+ *
+ * @template [V=unknown]
+ * @param {Iterable<[Key, V]>} rankSorted
+ * @param {RankCompare} rankCompare
+ * @param {FullCompare} fullCompare
+ * @returns {IterableIterator<[Key, V]>}
+ */
+const generateFullSortedEntries = (rankSorted, rankCompare, fullCompare) => {
```

Comment:
From the description, this sounds like a consolidation (or at least a relocation) of the functionality of `windowResort` and `bagWindowResort`. However, it does not replace them in this PR. Do we need all three? Or do I misunderstand the purpose of this one?

Reply:
The main difference seems to be that the old functions take arrays for their first argument, where this one takes an Iterable. But if so, it should be easy to replace the old ones with this one, yes?

Reply:
I intend to refactor the set and bag implementations in terms of these new functions and have added "_TODO share more code…_" comments to that effect, but to assist review of this PR have limited scope to just the compare functions `{set,bag,map}Compare = makeCompareCollection(…)`.

`makeCompareCollection` is the exposed foundation that is parameterized over type-specific entry generation and comparison functions, and it internally uses `generateCollectionPairEntries` to merge the entry sequences by key (which will also be needed in future exposed foundations like `makeIsSuperiorCollection`, `makeIsDisjointCollection`, `makeCollectionUnion`, `makeCollectionIntersection`, etc.) before comparing left and right sides of the returned paired entries. `generateCollectionPairEntries` itself internally relies upon `generateFullSortedEntries` to consistently resolve rank-order ties in the same way that merge-set-operators.js `merge` uses `windowResort` and merge-bag-operators.js `merge` uses `bagWindowResort`, and the `windowResort` and `bagResort` functions can indeed be replaced by the more general `generateFullSortedEntries`—but for that matter, so too can the `merge` functions be replaced by `generateCollectionPairEntries`, and so on up the call stack. But I didn't go that far here because it risks distracting from the main thrust of supporting CopyMaps in a narrow way, and instead stopped at the point necessary to show analogies in compareKeys.js.

Reply:
Ok cool. We cannot do everything first and this is a fine plan. Resolving.